### PR TITLE
Added Comments & intro/conclusion

### PR DESCRIPTION
**Pull request by Kacawi-zz**
- Added some documentation to the code chunks
- Added a begin and end to the tutorial with editorial note

### CODE REVIEW

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 1]
```diff
@@ -14,9 +17,13 @@
...
+    "Most of you are familiar with image data, taken with ordinary cameras (these are often called \"natural images\" in the scientific literature), but also with specialized instruments, such as microscopes or telescopes. When you're working with images in Python, you'll most often make use of the Matplotlib library, Python's 2D plotting library which produces publication quality figures in a variety of hard copy formats and interactive environments across platforms.\n",
+    "\n",
+    "Today's tutorial will show you how you can easily view 3D volumetric data with Matplotlib. \"3D volumetric data\" you say? Volumetric data is very common nowadays: it is captured by various technologies, e.g. MRI, CT, PET, USCT or echolocation. As you can already guess, this type of data plays an important role in medicine. \n",
     "\n",
-    "Using matplotlib to display such data is a straightforward call to the `imshow` function.\n",
+    "This tutorial will show you how to import and display image data, and how you can make a fully functional slice viewer to view all the slices in our MRI volume without any interference.\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: This is now quite repetitive ("today's tutorial will show you...", "this tutorial will show you..."). I'm also not a fan of using the standard matplotlib spiel to introduce it — it seems out of place to talk about plots in this context. Now that I see what you're going for, how about the following:

> Most of you are familiar with image data, taken with ordinary cameras (these are often called "natural images" in the scientific literature), but also with specialized instruments, such as microscopes or telescopes. When working with images in Python, the most common way to display them is using the `imshow` function of *Matplotlib*, Python's most popular plotting library.

> In this tutorial, we'll show you how to extend this function to display *3D volumetric data*, which you can think of as a stack of images. Together, they describe a 3D structure. For example, magnetic resonance imaging (MRI) and computed tomography (CT) scans measure the 3D structure inside the human body; X-ray microtomography measures the 3D structure inside materials such as glass, or metal alloys; and light-sheet microscopes measure fluorescent particles inside biological tissues. We'll demonstrate how to download an MRI dataset and display the slices using matplotlib. But first, let's go over some basics: how to display images with Matplotlib's `imshow`:

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 2]
```diff
@@ -54,7 +61,10 @@
    },
    "outputs": [],
    "source": [
+    "# import data module \n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: I would remove all comments about imports, as they almost repeat the line underneath them verbatim.

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 3]
```diff
@@ -54,7 +61,10 @@
...
     "from skimage import data\n",
+    "\n",
+    "# Gather some data \n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: This one is better, but I would rather place it in the markdown cell above it, as follows:

"""
Now we can import matplotlib and display some data. We'll load some example data that is included in the `data` module of the *scikit-image* library:
"""

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 4]
```diff
@@ -860,7 +870,10 @@
     }
    ],
    "source": [
+    "# Initialize the subplots\n",
     "fig, ax = plt.subplots(nrows=1, ncols=3)\n",
+    "\n",
+    "# Show subplot images and set titles \n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: These are good. =) Maybe reword as: "show an image in each subplot"

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 5]
```diff
@@ -860,7 +870,10 @@
     }
    ],
    "source": [
+    "# Initialize the subplots\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: "Initialize three subplot panels side by side"

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 6]
```diff
@@ -905,13 +921,19 @@
    },
    "outputs": [],
    "source": [
+    "# Import\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: Again, this particular type of comment, I think, adds nothing.

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 7]
```diff
@@ -975,17 +1012,25 @@
...
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
    },
    "outputs": [],
    "source": [
+    "# Import zipfile\n",
     "import zipfile\n",
+    "\n",
+    "# Extract the contents\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: Extract the contents into the temporary directory we created earlier

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 8]
```diff
@@ -1081,24 +1140,35 @@
...
+    "collapsed": true,
+    "deletable": true,
+    "editable": true
    },
    "outputs": [],
    "source": [
+    "# Read the image \n",
     "struct = nibabel.load(os.path.join(d, 'attention/structural/nsM00587_0002.hdr'))\n",
+    "\n",
+    "# Get a NumPy array\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: How about: Get a plain numpy array, without all the metadata

(but maybe that just makes it more complicated, I'm not sure)

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb [hunk 9]
```diff
@@ -4467,13 +4559,13 @@
...
     "\n",
-    "\n",
+    "# Go to the previous slice\n",
     "def previous_slice(ax):\n",
     "    volume = ax.volume\n",
     "    ax.index = (ax.index - 1) % volume.shape[0]  # wrap around using %\n",
     "    ax.images[0].set_array(volume[ax.index])\n",
     "\n",
-    "\n",
+    "# Go to the next slice\n",
```
Comment (jni) [inline notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: In this case (next to a def), I would prefer it as a function docstring:

```python
def next_slice(ax):
    """Go to the next slice."""
```

**Review by jni — CHANGES_REQUESTED** [review]
Hi @Kacawi! I've left a few comments. Let me know what you think and if you can, please update the PR! Thanks!

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb (hunk 1 above)
Comment (Kacawi-zz) [reply notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: @jni I understand. I agree with your proposal!

File: notebooks/Viewing 3D volumetric data with matplotlib.ipynb (hunk 2 above)
Comment (Kacawi-zz) [reply notebooks/Viewing 3D volumetric data with matplotlib.ipynb]: True. I'll remove them.

**Comment by Kacawi-zz** [comment]
@jni Thanks for the feedback! I made some changes in accordance with your feedback. Could you please review?

**Comment by jni** [comment]
Hi Karlijn! Annoyingly, github can't render the diff (looks like we've hit that limitation of using notebooks that we were worried about), but I went through it and have the following comments:

- it looks like `import matplotlib.pyplot as plt` is gone now! Specifically, this little section has some issues:

screen shot 2017-03-27 at 11 07 16 am

- the `urlretrieve` bit is a bit mangled also:

screen shot 2017-03-27 at 11 08 30 am

That's it! I like your new conclusion, it really wraps it up much more nicely than I had it. =) If you want, I can merge and fix those things myself, let me know.

**Comment by Kacawi-zz** [comment]
@jni I pushed some commits that should hopefully take care of the things you mention! Could you review if it's OK now?

**Comment by jni** [comment]
@Kacawi Github was still annoying but it helped actually, because I decided to just download your changes, make a couple of tweaks, and then merge directly in the command line. =)

The resulting notebook: https://github.com/jni/mpl-volume-viewer/blob/master/notebooks/Viewing%203D%20volumetric%20data%20with%20matplotlib.ipynb

**Comment by jni** [comment]
btw, for future reference, it's usually preferable to submit PRs from a different branch. =)
